test/maverick/consensus/misbehavior.go

```diff
@@ -0,0 +1,399 @@
+			return
+		}
+
+		// If ProposalBlock is nil, prevote nil.
+		if cs.ProposalBlock == nil {
+			logger.Info("enterPrevote: ProposalBlock is nil")
+			cs.signAddVote(tmproto.PrevoteType, nil, types.PartSetHeader{})
+			return
+		}
+
+		// Validate proposal block
+		err := cs.blockExec.ValidateBlock(cs.state, cs.ProposalBlock)
+		if err != nil {
+			// ProposalBlock is invalid, prevote nil.
+			logger.Error("enterPrevote: ProposalBlock is invalid", "err", err)
+			cs.signAddVote(tmproto.PrevoteType, nil, types.PartSetHeader{})
+			return
+		}
+
+		if cs.sw == nil {
+			logger.Error("nil switch")
+			return
+		}
+
+		prevote, err := cs.signVote(tmproto.PrevoteType, cs.ProposalBlock.Hash(), cs.ProposalBlockParts.Header())
+		if err != nil {
+			logger.Error("enterPrevote: Unable to sign block", "err", err)
+		}
+
+		nilPrevote, err := cs.signVote(tmproto.PrevoteType, nil, types.PartSetHeader{})
+		if err != nil {
+			logger.Error("enterPrevote: Unable to sign block", "err", err)
+		}
+
+		// add our own vote
+		cs.sendInternalMessage(msgInfo{&VoteMessage{prevote}, ""})
+
+		logger.Info("Sending conflicting votes")
+		peers := cs.sw.Peers().List()
+		// there has to be at least two other peers connected else this behavior works normally
+		for idx, peer := range peers {
+			if idx%2 == 0 { // sign the proposal block
+				peer.Send(VoteChannel, MustEncode(&VoteMessage{prevote}))
+			} else { // sign a nil block
+				peer.Send(VoteChannel, MustEncode(&VoteMessage{nilPrevote}))
+			}
+		}
+	}
+	return b
+}
+
+// DEFAULTS
+
+func DefaultEnterPropose(cs *State, height int64, round int32) {
+	logger := cs.Logger.With("height", height, "round", round)
+	// If we don't get the proposal and all block parts quick enough, enterPrevote
+	cs.scheduleTimeout(cs.config.Propose(round), height, round, cstypes.RoundStepPropose)
+
+	// Nothing more to do if we're not a validator
+	if cs.privValidator == nil {
+		logger.Debug("This node is not a validator")
+		return
+	}
+	logger.Debug("This node is a validator")
+
+	pubKey, err := cs.privValidator.GetPubKey()
+	if err != nil {
+		// If this node is a validator & proposer in the currentx round, it will
+		// miss the opportunity to create a block.
+		logger.Error("Error on retrival of pubkey", "err", err)
+		return
+	}
+	address := pubKey.Address()
+
+	// if not a validator, we're done
+	if !cs.Validators.HasAddress(address) {
+		logger.Debug("This node is not a validator", "addr", address, "vals", cs.Validators)
+		return
+	}
+
+	if cs.isProposer(address) {
+		logger.Info("enterPropose: Our turn to propose",
+			"proposer",
+			address,
+			"privValidator",
+			cs.privValidator)
+		cs.decideProposal(height, round)
+	} else {
+		logger.Info("enterPropose: Not our turn to propose",
+			"proposer",
+			cs.Validators.GetProposer().Address,
+			"privValidator",
+			cs.privValidator)
+	}
+}
+
+func DefaultEnterPrevote(cs *State, height int64, round int32) {
```

Comment:
can't we just use default funcs from consensus package? why do we need to copy them here?

Reply:
yeah but just for propose and prevote - there is no default pure function for precommit

Reply:
maybe it's something we can do in the future

Reply:
yeah I think it could be worth looking into